packages/@functionless/formation/src/resource-provider.ts

```diff
@@ -0,0 +1,140 @@
+// need to define how to create/update/delete a resource of type X when encountered.
+/**
+ * registration and discovery - For now, manually register.
+ *  Default deploy registers, collection of key to object with interfaces.
+ *   Can use object with more definitions.
+ * Separate promise for
+ */
+
+import { CloudControlProvider } from "./resource-providers/cloud-control";
+import { EventBusProvider } from "./resource-providers/event-bus";
+import { InlinePolicyProvider } from "./resource-providers/inline-policy";
+import { ManagedPolicyProvider } from "./resource-providers/managed-policy";
+import { QueuePolicyProvider } from "./resource-providers/queue-policy";
+import type { PhysicalResource, ResourceType } from "./resource";
+import { EventBusRuleProvider } from "./resource-providers/event-bridge-rule";
+import { RoleProvider } from "./resource-providers/role";
+import { QueueProvider } from "./resource-providers/queue";
+import { SecretProvider } from "./resource-providers/secret";
+import { NoOpProvider } from "./resource-providers/no-op";
+import { EventSourceMappingProvider } from "./resource-providers/event-source-mapping";
+
+export interface CreateRequest<Properties> {
+  logicalId: string;
+  resourceType: ResourceType;
+  definition: Properties;
+}
+
+export interface UpdateRequest<Properties> extends CreateRequest<Properties> {
+  previous: PhysicalResource<Properties>;
+}
+
+export interface DeleteRequest<Properties> {
+  logicalId: string;
+  resourceType: ResourceType;
+  physicalId: string;
+  previous: PhysicalResource<Properties>;
+  /**
+   * True when the resource was already snapshot before the delete.
+   */
+  snapshotDone: boolean;
+}
+
+export interface ResourceOperationResultMetadata {
+  /**
+   * Minimum milliseconds to wait after all deployment operations are complete.
+   *
+   * If it takes around 10s for a Policy update (paddingMillis: 10000) to be reflected (consistency), but
+   * the rest of the deployment only takes 5s, we will wait at least 5s before completing the deployment.
+   *
+   * TODO: replace with a consistent vs referable API.
+   */
+  paddingMillis?: number;
+}
+
+export type ResourceOperationResult<Properties = any> = Promise<
+  | PhysicalResource<Properties>
+  | ({
+      resource: PhysicalResource<Properties>;
+    } & ResourceOperationResultMetadata)
+>;
+
+export interface ResourceProviderRetryConfig {
+  canRetry: ("CREATE" | "DELETE" | "UPDATE")[] | true;
+}
+
+/**
+ * TODO: support optional snapshot.
```

Comment:
What do you mean

Reply:
Some resources support snapshots on delete. My thought was the interface would have an optional `snapshot` callback that can be implemented by a provider when they want to support snapshots.

Additionally, they should be able to force a snapshot on delete for cases like RDS.